alg/teca_spectral_filter.py

```diff
@@ -0,0 +1,253 @@
+    """
+    Filters data point-wise in the time domain using spectral analysis.
+    Enable spatial parallelism and set an appropriate window size.
+    """
+
+    def __init__(self):
+        """ initialize the class with defaults """
+
+        self.have_cuda = get_teca_has_cuda() and get_teca_has_cupy()
+        self.filter_type = 'low_pass'
+        self.filter_order = 2.
+        self.critical_frequency = 0.
+        self.point_arrays = []
+        self.pass_input_arrays = True
+
+    def set_filter_type(self, ftype, order):
+        """ set the filter type. The filter type can be low_pass or high_pass.
+        The order is typically a ordinal between 1 and 6.  A ValueError is
+        raised when unsupported filter type is encountered """
+
+        if ftype != 'low_pass' and ftype != 'high_pass':
+            raise ValueError('Invalid filter type %s. The filter type may be '
+                             'low_pass or high_pass') % (ftype)
+
+        if order < 1 or order > 6:
+            raise ValueError('The filter order should be between 1 and 6 but '
+                             'is %f' % (order))
+
+        self.filter_type = ftype
+        self.filter_order = order
+
+    def set_critical_period(self, period, units):
+        """ Set the filter's critical frequency from a period given in the
+        specified units. The units may be seconds, minutes, hours or days. This
+        could be used to filter on scales of familiar units. Eg. 3 days """
+
+        self.critical_frequency = 1. / ( period * self.get_number_of_seconds(units) )
+
+    def set_critical_frequency(self, freq):
+        """ Set the filter's critical frequency in units of Hz """
+
+        self.critical_frequency = freq
+
+    def set_pass_input_arrays(self, pass_in):
+        """ When set to True the input arrays will be included in the output.
+        In that case the filtered array will be named: <name>_<filter_type>
+        where name is the name of the input array and filter type is one of
+        low_pass or high_pass """
+
+        self.pass_input_arrays = pass_in
+
+    def set_point_arrays(self, arrays):
+        """ Set the list of arrays to filter """
+
+        if not isinstance(arrays, list):
+            arrays = list(arrays)
+
+        self.point_arrays = arrays
+
+    @staticmethod
+    def get_number_of_seconds(unit):
+        """ get the number of seconds in the specified unit. unit string must
+        contaoin one of seconds, minutes, hours, or days. A ValueError is
+        raised when invalid units are encountered """
+
+        if 'seconds' in unit:
+            return 1.
+        elif 'minutes' in unit:
+            return 60.
+        elif 'hours' in  unit:
+            return 60.*60.
+        elif 'days' in unit:
+            return 60.*60.*24.
+
+        raise ValueError('Unsupported time axis units %s' % (unit))
+
+    @staticmethod
+    def butterworth(npmod, n_samples, sample_rate, critical_freq, order, high_pass):
```

Comment:
I suggest not trying to implement the butterworth filter directly, rather take advantage of its implementation in scipy:  https://docs.scipy.org/doc/scipy/reference/generated/scipy.signal.butter.html

Reply:
I suppose I could have, however I found it overly complicated and wanted something that would easily translate to C++. Having had the experience of writing the simple version myself, I now feel more comfortable dealing with scpipy's complexity.

Reply:
this may be an easy change, I'll look into it.